### PR TITLE
139: Show only active categories

### PR DESCRIPTION
**Pull request by nuzil**
### Description

We fixed this issue with this PR
https://app.zenhub.com/workspace/o/magento/graphql-ce/issues/139
https://app.zenhub.com/workspace/o/magento/graphql-ce/issues/246

So currently only Active categories will be delivered with request

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=164) <br/>All committers have signed the CLA.

**Review by paliarush — APPROVED** [review]
Would be perfect if we can cover this with tests.

**Comment by naydav** [comment]
Pls check
```
Magento\GraphQl\Catalog\CategoryTest::testCategoriesTree
Failed asserting that null matches expected 'Ololo'.

/opt/bamboo/builds/MCCE23-AT1226-SOE/build-1/dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/CategoryTest.php:87
```

**Comment by andrewizotov** [comment]
> Pls check
> 
> ```
> Magento\GraphQl\Catalog\CategoryTest::testCategoriesTree
> Failed asserting that null matches expected 'Ololo'.
> 
> /opt/bamboo/builds/MCCE23-AT1226-SOE/build-1/dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/CategoryTest.php:87
> ```
Hello Valeriy , I tested default behaviour on develop "2.3-develop" branch  without my fix  . 
And I see some problems around child categories . please have a look . 
this is your categories tree . 
g1

Query from test 
image

and answer 
image

So category "Category 1.1.1" is nested in category ""Category 1.2" but this is not true . and "Category 12" doesn't have child nodes in tree, but in answer we have it . 
Could you investigate why it happen, seems to be all child categories always nested in last node of tree
Thanks

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php [hunk 1]
```diff
@@ -112,28 +112,7 @@ public function getTree(ResolveInfo $resolveInfo, int $rootCategoryId): \Iterato
...
         );
 
-        return $this->processTree($collection->getIterator());
-    }
-
-    /**
-     * @param \Iterator $iterator
-     * @return array
-     */
-    private function processTree(\Iterator $iterator) : array
```
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php]: Why is this code not necessary anymore?

Comment (nuzil) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php]: I guess it was mergin with 2.3-develop, cause currently I dont see processTree in https://github.com/magento/graphql-ce/blob/2.3-develop/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php [hunk 2]
```diff
@@ -42,14 +49,61 @@ public function execute(\Iterator $iterator): array
...
+                $tree = $currentLevelTree;
             }
+            $tree = $this->mergeCategoriesTrees($currentLevelTree, $tree);
         }
 
         return $tree;
     }
+
+    /**
+     * Merge together complex categories tree
```
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php]: How about `categories tree` => `category trees` ?
Same regarding method name itself.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php [hunk 3]
```diff
@@ -42,14 +49,61 @@ public function execute(\Iterator $iterator): array
...
+    }
+
+    /**
+     * Recursive method to generate tree for one category path
+     *
+     * @param $elements
+     * @param $index
+     * @return array
+     */
+    private function generateLevelTree($elements, $index): array
```
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php]: Method name is not reflecting its purpose, need to read the docs to understand what it is doing. Please consider improving method name.

**Review by paliarush — COMMENTED** [review]
This is pretty complex change, please make sure to cover it with GraphQL test.

**Comment by nuzil** [comment]
@paliarush Tests are there and aligned, just pushed last update

**Review by paliarush — APPROVED** [review]

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php [hunk 4]
```diff
@@ -42,14 +49,61 @@ public function execute(\Iterator $iterator): array
...
-            $nextCategory = $iterator->current();
-            $tree[$category->getId()] = $this->categoryHydrator->hydrateCategory($category);
-            $tree[$category->getId()]['model'] = $category;
-            if ($nextCategory && (int) $nextCategory->getLevel() !== (int) $category->getLevel()) {
-                $tree[$category->getId()]['children'] = $this->execute($iterator);
+
+            $pathElements = explode("/", $category->getPath());
+            $this->iteratingCategory = $category;
+
+            $currentLevelTree = $this->explodePathToArray($pathElements, self::START_CATEGORY_FETCH_LEVEL);
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php]: @paliarush @nuzil 
This changes will be broke current logic

Query:
```
{
  category(id: 334) {
   id
    name
  }
}
```

The result before changes:
```
{
  "data": {
    "category": {
      "id": 334,
      "name": "Cat-1"
    }
  }
}
```

The result after changes:
```
{
  "data": {
    "category": {
      "id": 2,
      "name": null
    }
  }
}
```
The root category has always been loaded first.

Looks like we need to change 
```
const START_CATEGORY_FETCH_LEVEL = 1;
```
on
```
const START_CATEGORY_FETCH_LEVEL = 2;
```

**Review by naydav — CHANGES_REQUESTED** [review]

**Comment by nuzil** [comment]
Hey @naydav, you was close. reason in different deepness level calculation for each category. I extended code and also added one additional test for this case

**Comment by raivisdejus** [comment]
@nuzil From the quick glance this fix is still missing sorting by the position. If you re-arrange the categories the tree in GraphQl will not match without this filter. That is implemented here https://github.com/magento/graphql-ce/pull/257/files#diff-b63924c5648c0a0b6c687d44f7bb7db3R105

@naydav  and @paliarush I think my solution for the ExtractDataFromCategoryTree.php is simpler and it also has unit tests to cover the basics of the implementation.

**Comment by nuzil** [comment]
@magento-engcom-team give me test instance

**Review by paliarush — APPROVED** [review]

**Comment by naydav** [comment]
We have static tests fail
1)
```
Magento\Test\Php\LiveCodeTest::testCodeMess
PHP Code Mess has found error(s):
/var/www/html/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php:24	The class CategoryTree has a coupling between objects value of 13. Consider to reduce the number of dependencies under 13.
```
2)
```
FILE: ...l/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 57 | ERROR | [x] Expected 1 space(s) after closing parenthesis;
    |       |     found 0
```

**Comment by nuzil** [comment]
Hi @naydav , I fixed test and removed dependency, let me know if there are some other changes required

**Pull request merged**